### PR TITLE
premerge: restore 256KB tx ceiling + scrub operator details from public repo

### PR DESCRIPTION
**Pull request by jsong1230**
Addresses the two merge-blocking findings from @trident90's #58 review. Both are properties of the dev tree, surfaced by promoting it to the public master release line.

## 1. txpool max-transaction regression (MEDIUM-HIGH)

The v1.13.14 rebase reverted `txMaxSize` to upstream `4*txSlotSize` (128KB) and deleted `params.MaxTransactionSize`, while Metadium keeps `MaxCodeSize=253952`. A creation tx for a near-max contract (~254–260KB) was rejected by the pool with `ErrOversizedData` **before** the EIP-3860 initcode check — a regression versus pre-Camellia mainnet (old master used `params.MaxTransactionSize`=262144). Restored via the named constant plus a pinning test (`TestTxMaxSizeCoversMaxCode`), so the next rebase fails loudly instead of silently reverting again.

**Scope**: this restores **pre-Camellia parity**, not full EIP-3860 reach — headroom over `MaxCodeSize` is 8KB, and consensus-legal initcode in (256KB, `MaxInitCodeSize`=2×`MaxCodeSize`] remains pool-rejected, as it was before Camellia. It is pool admission policy, **not consensus**: no chain-split risk, but nodes on an older binary reject and do not propagate 128–256KB transactions, so propagation of large deployments is path-dependent during a partial rollout — release-note entry drafted, runbook line added.

## 2. Operator details in a public repo (MEDIUM) — tree-scoped cleanup

- **Deleted**: `metadium/etcdutil.go.new` (1,395-line stale copy predating the etcd endpoint-validation security fix — restoring it would silently revert that fix), plus internal test-record docs (`docs/server36-*.md`, `docs/bench-net-tps-results.md`, `TODOS.md`) and superseded planning notes (`docs/plans/*`).
- **Scrubbed**: `tests/server36/scenarios/scenario-6-multiserver.sh` (env-driven `SSH_KEY`/`SERVERS`/`SERVER_NAMES`, test logic preserved), `scripts/update-node.sh`, `tests/private-net/setup.sh` + `Dockerfile.binary`, `tests/server36/Dockerfile.build-old` (now clones the official repo pinned to `m0.10.2.mainnet`), and the remaining personal paths in `docs/`.
- Removed the README link to the deleted TODOS.

**Scope — this cleanup is tree-only, not remediation of a disclosure.** The repo is public with ~21 forks: everything removed here remains retrievable from git history and from forks, and deleting it does not unpublish it. Exposure assessment (recorded in the review thread): no key material was ever committed — verified across full history — so the exposure is the key *filename* (`aws-jsong-nopass.pem`, advertising it is passphrase-less), operator account names, and RFC1918 internal addresses. **Rotation of the named fleet key is separate operational work; the decision is pending and nothing has been rotated yet.** This PR does not close that item.

## 3. `metclient.SendValue`: correctness fix, breaking for external importers

`amount` was a platform-sized `int` feeding `types.NewTransaction`'s wei value — silent truncation at ~2.1 gwei on 386 builds, and even `int64` caps at ~9.22 META. Now `*big.Int` (follow-on to #61/#62; full reasoning in the commit message of `e9a2d6c28`). No in-repo callers; external tooling that links `metclient` must update at compile time. Release-note entry drafted.

## Also (LOW, same review)

- `CLAUDE.md` + `scripts/update-node.sh`: corrected the stale `develop`/`feature/geth-v1.13.14` references to the actual `dev`→`master` flow (`update-node.sh` hard-failed on `git checkout develop` for any fresh clone).
- `.gitignore`: broadened the operator-local rule from `*.local.md` to `*.local.*`.

## Verification

- `go build` + txpool package build clean on the fleet toolchain; `TestTxMaxSizeCoversMaxCode` fails on a simulated revert to 128KB (run by @trident90 as well).
- CI unit_test (full suite incl. `core/txpool`) green at `15403f26f`; the one earlier red was the known `TestUpdatedKeyfileContents` keystore flake.
- Leak scan clean over the **working tree** (tree-scoped — see §2 for what remains in history).
- **Pending before the Aug 12 build**: one end-to-end ~254KB contract deployment on testnet (pool accept → propagate → block inclusion, and whether it fits under the block gas limit at all).

Merges into `dev`, then #58's transition branch is rebuilt on top so its tree stays bit-identical to dev — #58 will get a fresh verification comment and the existing one will be marked superseded.

Follow-ups from the same review (non-blocking, post-activation): #63 (i386/full-suite coverage), #65 (miner deadline-guard/prefetcher, deploy-tool `return nil`, CI dedup). Remaining review nits (test headroom constant, `params` non-consensus marker line, `GMET_BIN` guard, scenario-6 bash-4 prereq note, `*.local` extensionless glob) are deferred to a follow-up unless a reviewer wants them in this PR.

### CODE REVIEW

**Review by cp-khs — COMMENTED** [review]
Thanks for turning these around fast, and for requesting reviewers before merging this time. The `SendValue` commit is exactly right — the message explaining *why* the i386 job couldn't catch it is the part someone will thank you for in a year. trident90's txpool find is a good catch too.

One finding below is bigger than the PR body suggests, and it isn't something this PR can fix.

---

## 1. Deleting and scrubbing does not remove any of this from history

This repo is public and has **21 forks**. Everything removed in `cf8cd9c` is still fully retrievable:

```
git log -p -- tests/server36/scenarios/scenario-6-multiserver.sh
git show <pre-scrub-sha>:metadium/etcdutil.go.new
```

Old blobs are browsable on github.com by SHA, they persist in every fork, and GitHub serves unreachable commits by SHA for a long time. Rewriting history isn't realistic here — #58 is about to promote this tree to `master` and the forks are outside our control.

So this PR is **hygiene going forward, not remediation of a disclosure.** The only real response is to treat the contents as already public:

- **SSH key** — if the filename maps to a key actually deployed on the fleet, rotate it. Worth confirming whether only the filename was exposed or the path/identity ties back to a live key.
- **Operator account names** — not credentials, but they are targeting information. If those names are real SSH login or sudo accounts, factor that into the exposure assessment.
- **Internal IPs** — can't be rotated. Worth checking that no firewall or access policy is relying on those addresses being unknown.

The body says "leak scan of the tree is clean," which is literally accurate — *tree*, not history. But as written it will be read internally as "this is closed." Could you state plainly in the body that the content remains in history and forks, and record what was rotated? If nothing has been rotated yet, that's the actual open work here, and it may outrank the Aug 12 build.

Same point applies to `etcdutil.go.new`: a 1,395-line copy predating the etcd endpoint-validation fix means the pre-fix code has been sitting in a public repo. Deleting it now doesn't unpublish it. If that fix addressed something exploitable, that deserves its own assessment rather than riding along in a hygiene commit.

## 2. Bring back the named constant, and add a test that fails on the next rebase

`8*txSlotSize` = 262144, matching the pre-Camellia value exactly. The restoration is correct.

But look at *how* this regression happened: a rebase silently reverted a fork-specific value to the upstream default and nobody noticed until the tree was promoted to a public release line. `8*txSlotSize` is indistinguishable from ordinary upstream code, so **the next go-ethereum rebase will silently revert it the same way.**

- Restore `params.MaxTransactionSize` and use it here, with a comment tying it to `MaxCodeSize`. A grep-able fork-specific name is what makes it visible during a rebase.
- Add a test asserting the relationship (something like `txMaxSize >= params.MaxCodeSize` plus headroom). The next rebase should **fail loudly** rather than quietly shipping 128KB again.

Without that, this fixes the symptom and leaves the mechanism intact.

Minor related note: EIP-3860's `MaxInitCodeSize` (normally `2*MaxCodeSize`) sits above 256KB, so initcode between 256KB and that bound is still rejected by the pool after this change. Restoring pre-Camellia parity is the right goal — just worth a line in the body so nobody reads this as "3860 is now fully reachable."

## 3. "build clean" doesn't verify a runtime policy change

The stated verification is `go build` plus a package build. `txMaxSize` is a runtime admission policy — a successful build proves nothing about it. Before this goes into the Aug 12 binary:

- `go test ./core/txpool/...` (legacypool has size-related cases)
- one **actual ~254KB contract deployment on testnet**, end to end: accepted by the pool, propagated, included in a block. That also settles whether it fits under the block gas limit at all, which no amount of reasoning from constants will tell you.

Two days before the public build, one empirical data point is worth more than the derivation.

Also worth a runbook line: `txMaxSize` is local pool policy, not consensus, so this won't split the chain — but nodes still on the old binary will **reject and refuse to propagate** a 256KB transaction. During the Aug 20–27 rolling window, and indefinitely for external nodes that never upgrade, propagation of large deployment transactions is partial.

## 4. This invalidates #58's verification comment

Merging here moves the `dev` tip off `c45b5e716`, so when #58's branch is rebuilt its tree hash, both parents, and all four CI results change. You've clearly accounted for this ("rebuilt on top so its tree stays bit-identical to dev"), but #58 will need a **fresh verification comment** with the new hashes, and the existing one should be collapsed or marked superseded so reviewers don't check against stale values.

## 5. Split the txpool change into its own commit

`cf8cd9c` currently bundles a pool policy change, security scrubbing, file deletions, doc fixes, and a gitignore tweak. For something landing on a release line days before a fork, the txpool change should stand alone so it can be reverted or cherry-picked independently and so `git log -- core/txpool/` stays readable.

## Looks good

- `SendValue` amount widening, with the reasoning about silent truncation preserved in the commit message
- `update-node.sh` — `git checkout develop` hard-failing on any fresh clone was a real bug, not just a stale reference
- splitting the non-blocking items out to #63 and #65 rather than letting them ride along

---

2, 3 and 5 are all doable within this PR and shouldn't cost more than half a day. **1 is not something this PR can close** — the exposure assessment and any rotation are separate work, and if that hasn't happened yet I'd rank it above the Aug 12 build.

**Comment by jsong1230** [comment]
@cp-khs Excellent review — all five addressed. The head is now four separate commits (`70c852f` txpool / `207d6cd` metclient / `590947e` scrub+deletions / `11e47fa` housekeeping).

**#2 — named constant + failing test.** Done (`70c852f`). Restored `params.MaxTransactionSize = 262144` and pointed `txMaxSize` at it instead of `8*txSlotSize`, with a comment tying it to `MaxCodeSize` and explaining the rebase-revert history. Added `TestTxMaxSizeCoversMaxCode` (`txMaxSize >= MaxCodeSize` and `== MaxTransactionSize`). **Verified the guard bites**: reverting to `4*txSlotSize` makes the test FAIL, so the next rebase that silently reverts the ceiling breaks the build instead of shipping 128KB. Added the EIP-3860 note you asked for (initcode between 256KB and `2*MaxCodeSize` stays pool-rejected; this only restores pre-Camellia parity) and the 'local admission policy, not consensus' line.

**#3 — runtime verification.** `go test ./core/txpool/legacypool/` passes (size cases + the new guard). The **end-to-end ~254KB deployment on testnet** is the one piece I haven't done yet — agreed it's the empirical data point that matters, and I'll run it (accept → propagate → include in a block, and confirm it fits under the block gas limit) before the Aug 12 build rather than reasoning from constants. Will post the result here.

**#4 — #58 verification comment.** Correct, merging here moves the dev tip. When I rebuild #58's branch I'll post a fresh verification comment with the new hash/parents/CI and mark the current one superseded so nobody checks stale values.

**#5 — split commits.** Done — the txpool policy change is `70c852f` alone, so it can be reverted or cherry-picked independently and `git log -- core/txpool/` stays readable.

**#1 — this is the one that isn't a code fix, and you're right that it outranks the build.** Confirmed factually: **no key material was ever committed** — no `.pem` blob, no `BEGIN … PRIVATE KEY` anywhere in history; what leaked is the *filename* (`aws-jsong-nopass.pem`, which advertises passphrase-less), three operator account names, and internal 10.150.255.x addresses. So this is targeting/metadata exposure in a public repo with 21 forks, permanent in history — not a private-key disclosure. Rotation of that key and the exposure assessment are an operations decision, not something this PR closes; I've escalated it to the owner as separate work ranked ahead of the build, and the commit body now says plainly that deletion is forward hygiene and the content persists in history/forks. I'll link the tracking item once it's filed.

Re-requesting review at `11e47fa`.

**Comment by jsong1230** [comment]
Re **#1 (key/detail exposure)** — decision recorded, so this isn't left as silence:

- **Assessed**: no key material was ever committed (verified across full history — no PEM blob, no `BEGIN … PRIVATE KEY`). Exposed content is the *filename* (`aws-jsong-nopass.pem`, advertising passphrase-less), three operator account names, and internal 10.150.255.x (RFC1918, not externally routable) addresses — targeting metadata, not a credential.
- **Decision**: rotate the fleet SSH key as **planned post-activation work**, folded into the operator-offboarding key cleanup already on the books, not as an emergency. Rationale: no private-key disclosure, the IPs are private, and rotation is a 20-node `authorized_keys` operation better done deliberately than under release pressure. The exposure is now on record as permanent-in-history (the commit body says so), so nothing here reads as 'closed'.
- The internal IP / firewall-reliance check you flagged is part of that same assessment.

This is operational and independent of the code in this PR — #64 can proceed on its own merits.

File: core/txpool/legacypool/legacypool.go [hunk 1]
```diff
@@ -51,7 +51,13 @@ const (
...
 	// more expensive to propagate; larger transactions also take more resources
 	// to validate whether they fit into the pool or not.
-	txMaxSize = 4 * txSlotSize // 128KB
+	//
+	// Metadium pins this to params.MaxTransactionSize (256KB) rather than the
+	// upstream 4*txSlotSize (128KB): a creation tx for a max-size contract
+	// (params.MaxCodeSize) runs ~256KB and must be admitted, not rejected with
+	// ErrOversizedData before the EIP-3860 check. The named constant is
+	// deliberate -- see the note on params.MaxTransactionSize.
+	txMaxSize = params.MaxTransactionSize
```
Comment (trident90) [inline core/txpool/legacypool/legacypool.go:61]: **[MEDIUM — partial fix + rebase-fragile shape]** Two issues with the restore:

1. **The comment contradicts its own arithmetic.** `8 * txSlotSize = 262144` exactly, and `validation.go:62` rejects on strict `>`, so a tx that "runs just over 256KB" (this comment's words) is still rejected by this very constant. Real headroom over `MaxCodeSize` (253952) is 8192 bytes — a max-size deployment with more than ~8KB of ABI-encoded constructor args still gets `ErrOversizedData`. And consensus-legal initcode in (262144, 507904] (`MaxInitCodeSize = 2*MaxCodeSize`) is still pool-rejected before the EIP-3860 check ever runs — the failure mode is narrowed, not eliminated. Either fix the comment to "just under 256KB" and document the 8KB constructor-arg ceiling, or size the constant off `MaxInitCodeSize` if the intent is to make everything consensus-legal submittable.

2. **The value is re-added as an inline multiplier in an upstream-owned file instead of restoring `params.MaxTransactionSize`** (old master's pattern: `params/protocol_params.go` constant consumed by the pool). This inline-edit shape is exactly what the v1.13.14 rebase silently reverted last time — and `legacypool_test.go` derives all its bounds *from* `txMaxSize`, so no test pins 262144: the next rebase that reverts to `4*txSlotSize` keeps CI green while re-breaking max-size deployments. `legacypool.go` already imports `params`; restoring the named constant plus a one-line test that asserts 262144 makes the fork delta rebase-survivable.

File: tests/server36/scenarios/scenario-6-multiserver.sh [hunk 2]
```diff
@@ -1,27 +1,36 @@
...
 
-SSH_KEY="${SSH_KEY:-$HOME/.ssh/aws-jsong-nopass.pem}"
-SERVERS=(
-  "cplabs@10.150.255.36"
-  "jsong@10.150.255.31"
-  "jsuhr@10.150.255.33"
-)
-SERVER_NAMES=(server36 server31 server33)
+SSH_KEY="${SSH_KEY:?set SSH_KEY to the private key for the benchmark hosts}"
+read -r -a SERVERS <<< "${SERVERS:?set SERVERS to a space-separated user@host list}"
```
Comment (trident90) [inline tests/server36/scenarios/scenario-6-multiserver.sh:27]: **[MEDIUM — the variable-length refactor didn't reach the rest of the script]** `SERVERS` is now an arbitrary-length env-driven array, but four downstream spots still assume the old fixed 3-host setup (the first was **reproduced by execution**):

1. **Line 64**: the launch loop is still `for i in 0 1 2`. With 2 hosts, the `i=2` iteration expands `${SERVERS[$i]}` inside the backgrounded subshell, which dies with "unbound variable" under `set -u` — but `wait "$pid" || true` (line 71) swallows it and the script **exits 0**. With 4+ hosts, indices 3..N are never launched, yet the report generator receives all `SERVER_NAMES`, producing phantom all-N/A columns after a multi-hour benchmark. Fix: `for i in "${!SERVERS[@]}"`.
2. **Line 153**: `write_summary "$RESULT_DIR" 1 0 ...` hardcodes pass=1/fail=0, so the scenario reports PASS even when every remote run failed (every failure path is discarded via `|| true`). Combined with (1), a 2-host invocation runs zero benchmarks and still writes a PASS summary. Sibling scenarios pass computed counts (`scenario-6.sh:414`).
3. **Line 147**: the report heredoc is opened with a quoted delimiter (`<<'PYEOF'`), so `$(date -u ...)` on this line is emitted **literally** into combined-comparison.md — python f-strings don't expand shell syntax.
4. **Lines 145-146**: the template still asserts "all servers are identical hardware: Intel i5-11400T 12-core, 15GB RAM" — now an unverifiable claim about arbitrary env-supplied hosts, inviting wrong conclusions from benchmark variance.

File: tests/server36/scenarios/scenario-6-multiserver.sh [hunk 3]
```diff
@@ -1,27 +1,36 @@
...
+read -r -a SERVER_NAMES <<< "${SERVER_NAMES:-}"
+if [ "${#SERVER_NAMES[@]}" -ne "${#SERVERS[@]}" ]; then
```
Comment (trident90) [inline tests/server36/scenarios/scenario-6-multiserver.sh:33]: **[LOW]** Two edge cases in the name derivation:
- A user-supplied `SERVER_NAMES` whose length doesn't match `SERVERS` is **silently replaced** with derived names — worth at least a warning to stderr, since the user's labels vanish from the report without explanation.
- Derived names collide for same-host entries: `alice@h1` and `bob@h1` both derive `h1`, so the remote results dir, the local `$RESULT_DIR/h1-bench.log`, and the scp target `$RESULT_DIR/h1/` (lines 47/53/58) clobber each other and the report shows one merged column. Suffixing with the array index on collision fixes it.

File: scripts/update-node.sh [hunk 4]
```diff
@@ -40,10 +40,10 @@ log "Repository: $REPO_DIR"
 
 # Pull latest code
 cd "$REPO_DIR"
-log "git pull origin develop..."
+log "git pull origin dev..."
```
Comment (trident90) [inline scripts/update-node.sh:43]: **[MEDIUM-LOW — scrub incomplete]** The operator-details scrub misses several spots, including this very file: lines 8 and 87-88 still carry `/data/jsong/...` paths. Elsewhere:
- `tests/private-net/setup.sh:11` and `Dockerfile.binary:7` — `/data/jsong/gmet-rocksdb`
- `tests/server36/Dockerfile.build-old:15-17` — clones the personal fork `jsong1230/go-metadium` at `FEATURE_BRANCH=feature/geth-v1.13.14`, the branch this PR's CLAUDE.md edit just removed from the documented strategy, so the old-binary baseline build now depends on an undocumented branch of a personal repo
- `docs/system/system-analysis.md:581-586`, `docs/camellia-verification-checklist.md:270`, `docs/triedb-alignment-summary.md:123` — personal `/Users/jsong` paths and the `gram-jsong` hostname

Since the PR's stated goal is making the tree consistent with CLAUDE.md's "operational details are intentionally kept out" policy, these should ride along in the same pass.

File: metadium/metclient/util.go [hunk 5]
```diff
@@ -436,7 +436,7 @@ func SendContract(ctx context.Context, contract *RemoteContract, method string,
 	return
 }
 
-func SendValue(ctx context.Context, cli *ethclient.Client, from *keystore.Key, to common.Address, amount, gas int, _gasPrice int64) (hash common.Hash, err error) {
+func SendValue(ctx context.Context, cli *ethclient.Client, from *keystore.Key, to common.Address, amount int64, gas int, _gasPrice int64) (hash common.Hash, err error) {
```
Comment (trident90) [inline metadium/metclient/util.go]: **[LOW]** `int64` is better than `int`, but it doesn't fix the range problem the type change is for: `amount` is wei (line 450 feeds it into `types.NewTransaction`'s value), and `math.MaxInt64` ≈ 9.223e18 wei ≈ **9.22 META** — an ordinary 100 META transfer (1e20 wei) still can't be expressed; the caller overflows before `big.NewInt` is ever reached. There are zero in-tree callers (grep: definition + an unrelated 4byte.json entry), so the signature is already being broken for external callers by this very change — `amount *big.Int` costs the same migration and actually removes the ceiling.

**Review by trident90 — COMMENTED** [review]
Follow-up review for the PR #58 pre-merge items. Candidates were collected from eight review angles and verified (the shell findings were reproduced by actually executing the script).

**What checks out cleanly**: `etcdutil.go.new` deletion (nothing references it; the live `etcdutil.go` retains the endpoint-validation fix), `TODOS.md` removal (no dangling links — the README link removed here was the only one), the `develop` → `dev` CLAUDE.md/scripts fix (`origin/dev` exists, no leftover references in fork-owned files), and the `.gitignore` widening (no tracked file matches, no rule overlap).

**Where the fixes fall short** (details inline):
- The 256KB restore is partial — the new comment contradicts its own arithmetic, real headroom over `MaxCodeSize` is only 8KB, and the constant is inlined in an upstream-owned file with no pinning test, which is exactly the shape the last rebase silently reverted.
- `SendValue`'s `amount` went `int` → `int64`, but int64 wei still caps at ~9.22 META; `*big.Int` costs the same signature break and actually fixes the range.
- The operator-details scrub misses several files, including lines in a file this PR itself edits.
- The scenario-6 script's env-driven `SERVERS` refactor left the launch loop, the PASS/FAIL summary, and the report template un-adapted (launch-loop failure reproduced by execution).

**Verified and not flagged**: p2p propagation of 256KB txs is fine (both `maxTxRetrievalSize` and `maxTxPacketSize` accumulators admit one tx before testing the byte budget, so a max-size tx is still requested/sent alone; `txMaxBroadcastSize=4096` already forces announce-only), and the CLAUDE.md "Project Management" staleness candidate was refuted.

**Comment by jsong1230** [comment]
@trident90 Thorough — all four addressed at `06ed50c`. (Note: the named-constant + pinning-test part of finding 1 was already in the tree at review time; the arithmetic sub-point and everything else were real and are fixed.)

**1. txpool — comment arithmetic + rebase-fragility.**
- Comment corrected: the 8KB headroom over `MaxCodeSize` is now stated explicitly, along with the fact that a max-size deployment with >~8KB constructor args, and consensus-legal initcode in (256KB, 2·MaxCodeSize], remain pool-rejected. Framed as pre-Camellia parity, not full EIP-3860 reach — kept at 262144 rather than sizing off `MaxInitCodeSize`, since raising it widens propagation cost beyond historical behaviour.
- `txMaxSize = params.MaxTransactionSize` (named constant restored in `params/protocol_params.go`), and `TestTxMaxSizeCoversMaxCode` asserts `txMaxSize >= MaxCodeSize`. **Verified the guard bites**: reverting to `4*txSlotSize` fails the test.

**2. scenario-6 — the refactor didn't reach the rest of the script.** Fixed all four, plus the two edge cases from your LOW note:
- launch loop → `for i in "${!SERVERS[@]}"`, and `run_remote_bench` now returns the remote rc so failures aren't swallowed
- PASS/FAIL computed from `wait` results, passed to `write_summary` (no more hardcoded 1/0)
- report timestamp computed in shell and passed via `REPORT_TS` env, not `$(...)` inside the quoted heredoc
- dropped the identical-hardware claim for arbitrary env hosts
- `SERVER_NAMES` length mismatch now warns to stderr instead of silently replacing; derived-name collisions get an index suffix

**3. scrub incomplete.** Completed across everything you listed and a full-tree re-grep is now clean: `update-node.sh` (lines 8/87-88), `tests/private-net/setup.sh`, `Dockerfile.binary`, `docs/system/system-analysis.md`, `docs/triedb-alignment-summary.md`, `docs/camellia-test-report.md`. `Dockerfile.build-old` now builds from `METADIUM/go-metadium` at the `m0.10.2.mainnet` tag + `dev` instead of the personal fork at the removed `feature/geth-v1.13.14` branch — which also gives that tag a concrete consumer.

**4. `amount` → `*big.Int`.** Taken — `int64` still capped at ~9.22 META; `*big.Int` removes the ceiling for the same signature break, zero in-tree callers.

Thanks also for verifying the 256KB p2p propagation path (retrieval/packet accumulators admit one tx before the byte-budget test) — that's the reassuring half of the txpool change. Re-requesting at `06ed50c`.

**Review by trident90 — COMMENTED** [review]
Re-review of 06ed50c — **all four review items verified as addressed**, with one small leftover.

**1. txpool ceiling — verified.** The `params.MaxTransactionSize` comment now documents the exact scope: 8KB headroom over `MaxCodeSize`, the (256KB, `MaxInitCodeSize`] gap, and the pre-Camellia-parity (not full-EIP-3860-coverage) intent. I ran `TestTxMaxSizeCoversMaxCode` — it passes, and its first assertion (`txMaxSize < params.MaxCodeSize`) is what trips if a rebase reverts the ceiling to 128KB, so the guard genuinely fails loudly on the regression it was written for.

**2. scenario-6 — verified, including the LOW item.** Launch loop is `for i in "${!SERVERS[@]}"`, per-host rc is captured and tallied into a real PASS/FAIL passed to `write_summary` with `[ "$FAIL" -eq 0 ]` as the exit status, `REPORT_TS` is computed outside the quoted heredoc and read via `os.environ`, the hardware assertion is neutralized, and SERVER_NAMES now warns on length mismatch and suffixes the index on same-host collisions. `bash -n` passes. One portability nit, take it or leave it: `declare -A` requires bash 4+, so the driver won't run under macOS's stock bash 3.2 — irrelevant if this is always launched from Linux.

**3. Operator-details scrub — one leftover.** update-node.sh, private-net setup.sh/Dockerfile.binary, and the docs are clean, and Dockerfile.build-old now clones the official repo pinned to `m0.10.2.mainnet` (I confirmed the tag exists on origin and points at the 0.10.2 line) with ABI files from `dev` — a strictly better baseline than the personal-fork branch. However, **`docs/camellia-verification-checklist.md:270` still carries the personal path** `~/.gstack/projects/jsong1230-go-metadium/jsong-feature-camellia-design-20260403-235011.md` — this was the third docs file in the original comment (camellia-test-report.md was fixed instead). A full-tree grep shows it is the only personal-detail hit remaining.

**4. `SendValue` — verified.** `amount *big.Int` passed straight through to `types.NewTransaction`; no in-tree callers to migrate; `go build ./metadium/... ./core/txpool/...` passes.

With that one checklist.md line scrubbed, no objections to merge from my side.

**Review by cp-khs — COMMENTED** [review]
PR #64 — lean review (post-ready)

Two parts. Part 1 goes in the review body box, Part 2 as line-anchored inline comments. Nothing else.

PART 1 — Review body (select "Approve", paste this)

Approving. I checked the regression claim independently rather than taking it from the description, and it holds: MaxTransactionSize = 262144 is present at origin/master:params/protocol_params.go:127, and master:core/tx_pool.go:59 was already wired to it, so the Camellia rebase genuinely dropped this and the PR restores prior behaviour rather than setting new policy. I also went through the deleted metadium/etcdutil.go.new line by line — it's missing the entire etcd cluster-membership validation fix and its etcdAutoJoinLock is a nil channel, so anyone who copied it over the live file would reintroduce a membership-hijack path plus a deadlock. Deleting it is the right call, more clearly so than the commit message suggests.

One caveat on my sign-off: I couldn't compile or run tests in my environment, so this is source-level review only. Please confirm CI is green — I'm not vouching for build status.

Three things I'd like resolved or explicitly deferred before this lands, none of them blocking:

The scrub is tree-only. This is a public repo and history still has the internal addresses, the operator accounts, and aws-jsong-nopass.pem — a filename that advertises the key has no passphrase. I don't think a history rewrite is warranted here since no key material was ever committed and the addresses are RFC1918, but the description's "scrub from the public repo" claims more than the diff delivers. Suggest rotating that key regardless, since its name is public either way, and adjusting the description to say the cleanup is tree-scoped.

SendValue needs a better commit message and a release note. The change reads as cleanup but it's a correctness fix: amount int capped transfers at about 2.1 gwei on 386 builds and at int64 elsewhere. Worth stating, especially as a follow-on to #61/#62. It's also a breaking change to an exported API — no in-repo callers, but external tooling links metclient.

The two comments on the same guarantee disagree. legacypool.go says a max-size deployment "must be admitted"; params correctly says this "does not make every consensus-legal deployment submittable." The params version is right — MaxCodeSize bounds deployed runtime code, the transaction carries initcode, and 8KB of headroom only covers a near-trivial constructor. Left as-is, the strong version will get cited as a guarantee. Details inline.

Rest is inline and non-blocking. Separately, the three fixes you made outside the stated scope are the most valuable part of the hygiene commits: the $(date) that was being emitted literally into every combined report, the summary that wrote PASS when every host's benchmark had failed, and the result-directory collision between two accounts on one host. Also good call moving Dockerfile.build-old off the personal fork onto a pinned tag — the reasoning in the header about the upcoming master transition will save someone real time later.

PART 2 — Inline comments
core/txpool/legacypool/legacypool.go line 56

This is stronger than the constant guarantees, and it contradicts the qualified version you wrote in params. MaxCodeSize bounds deployed runtime code; the tx carries initcode, which is necessarily larger (constructor, ABI args, metadata) and is bounded by MaxInitCodeSize = 2 * MaxCodeSize. With 8192 bytes of headroom a max-size deployment fits only with a near-trivial constructor. Suggest matching the params wording — something like "a max-size contract deployment with a small constructor is admitted rather than rejected with ErrOversizedData before the EIP-3860 check; see params.MaxTransactionSize for the exact scope."

core/txpool/legacypool/txmaxsize_metadium_test.go line 20

The invariant you're defending is txMaxSize >= MaxCodeSize + creation overhead, but this passes with zero headroom — exactly the state that breaks max-size deployments. Making it explicit puts the intent in the code:

go
// Constructor bytecode, ABI-encoded arguments, RLP envelope, signature.
const minCreationOverhead = 8192

if txMaxSize < params.MaxCodeSize+minCreationOverhead {

The guard's overall shape is good, though — if a rebase drops MaxTransactionSize from params entirely (what happened last time) legacypool.go won't compile, and if it only reverts the value this test fails. Both paths are loud.

params/protocol_params.go line 152

This is pool admission policy, not a consensus parameter, but it now sits between MaxCodeSize and the precompile gas costs. Keeping it here is right for parity with master, but someone scanning this block will assume changing it needs a hard fork. One line would fix it: not consensus-critical -- pool admission policy only.

.gitignore line 73

*.local.* misses extensionless foo.local. Suggest adding *.local too. Heads-up in the other direction as well: the broader glob will silently ignore intentionally-committed files like config.local.json. Nothing in the tree matches today, so no action needed now.

tests/private-net/setup.sh line 11

scenario-6-multiserver.sh in this same PR uses :? to fail fast with an actionable message; here the placeholder default defers it to a confusing no such file or directory. Suggest GMET_BIN="${GMET_BIN:?set GMET_BIN to the gmet binary}" for consistency.

tests/server36/scenarios/scenario-6-multiserver.sh line 32

declare -A needs bash 4+, but the shebang resolves to 3.2 on stock macOS. Fine if this is server-only — worth a line in the prerequisites block.

Also: SERVERS=" " passes the :? guard and leaves PIDS empty, and "${PIDS[@]}" under set -u is an unbound-variable error on bash <= 4.3. [ "${#SERVERS[@]}" -gt 0 ] || { log "SERVERS is empty"; exit 2; } after the guard closes it.

Minor: the collision-suffix logic only runs on the derived-names path, so user-supplied SERVER_NAMES with duplicates still clobber each other — only the count is checked.

params/protocol_params.go line 152 (second comment, or raise in the release thread)

Two release-note entries needed: the pool ceiling change (no consensus impact, but un-upgraded nodes still reject above 128KB, so propagation of 128-256KB transactions is path-dependent during a partial rollout), and the SendValue signature break.

Worth confirming separately, outside this PR: depositing 253,952 bytes of code costs roughly 50.8M gas for the code deposit alone, before calldata. If the block gas limit doesn't accommodate that, the restored headroom is theoretical — which is still fine, parity with pre-Camellia stands on its own, but we should know which it is.

**Comment by jsong1230** [comment]
@trident90 The leftover is scrubbed at `15403f26f` — `docs/camellia-verification-checklist.md:270` referenced a session-local design note outside the repository, so there was no repo-relative path to substitute; the line is dropped. Full-tree grep is now clean (the remaining `.gstack` mention in `camellia-test-report.md:95` is the fix-history entry describing the gitignore rule itself, not a personal path — matching your grep result). On the `declare -A` portability nit: leaving as-is, the driver only ever runs from the Linux bench hosts.

@cp-khs Thanks — the three items are resolved or explicitly recorded, at `9f604d0c9` + an updated PR description:

1. **Tree-scoped scrub** — the description now states plainly that the cleanup is tree-only, that history and the ~21 forks retain everything, and that **nothing has been rotated yet** (key rotation is tracked as separate operational work, decision pending). The "leak scan clean" claim is now explicitly tree-scoped.
2. **SendValue** — the commit message of `e9a2d6c28` already carries the full reasoning (386 ~2.1 gwei truncation, int64 ~9.22 META cap, breaking change for external importers at compile time); the PR description now surfaces it as its own section, and both release-note entries are drafted (pool-ceiling propagation during partial rollout + the signature break).
3. **Comment contradiction** — `legacypool.go` no longer says "must be admitted"; it now matches the qualified `params` wording (`9f604d0c9`): a max-size deployment *with a small constructor* is admitted, exact scope documented at `params.MaxTransactionSize`.

CI status you asked to confirm: unit_test is green at `15403f26f` (full suite; the earlier red was the known `TestUpdatedKeyfileContents` keystore flake, third occurrence), and `9f604d0c9` (comment-only) is running now. The remaining empirical check — an end-to-end ~254KB deployment on testnet incl. block-gas-limit fit — is scheduled before the Aug 12 build, per your item 3.

Your remaining inline nits (test headroom constant, `params` non-consensus marker, `GMET_BIN` guard, scenario-6 bash-4 prereq, `*.local` glob) are deferred to a follow-up unless you want them in this PR.

One housekeeping note: your second review reads as an approval but was submitted as a comment (the draft's paste instructions are included verbatim at the top). When you're satisfied, could you re-submit via the **Approve** state — the merge policy here counts the formal review state.

**Review by trident90 — APPROVED** [review]
LGTM

**Review by cp-khs — APPROVED** [review]
LGTM
